### DISCUSSION TITLE
Correlated CTEs in subquery

### DISCUSSION THREAD

**Discussion by paladam1** [staudenmeir/laravel-adjacency-list#223] [Q&A]
Hi,

I need to use the same as the example, `whereHas('descendant')` on a MariaDB.
Is there a performance-friendly workaround for that issue?

Thanks for your help.

**Comment by staudenmeir** [comment]
Hi @paladam1,
What does your whole `whereHas('descendants')` call look like?

**Reply by paladam1** [reply]
Hi @staudenmeir,

thank you for the quick reply.

My query is the following:

```
Project::tree()
            ->with([... some not recursive relations ...])
            ->withExists([
                'children',
            ])
            ->where(function ($query) {
                $query->search(['name' => $this->name, ... and so on]);
                $query->when($this->status === 0, fn(Builder $q) => $this->filterStatus0($q));
                $query->when($this->status === 1, fn(Builder $q) => $this->filterStatus1($q));
            })
            ->orWhereHas('descendants', function ($query) {
                // Query the same as the above where.
            })
            ->get();
```

**Reply by paladam1** [reply]
I was just thinking about using `withQueryConstraint` because I'm trying to get all results excluding some elements and their subtrees.

**Comment by staudenmeir** [comment]
There is no good workaround, AFAIK. Does your tree have a maximum depth? If so, you can use this workaround:

```php
$constraint = function ($query) {
    $query->search(['name' => $this->name, ... and so on]);
                $query->when($this->status === 0, fn(Builder $q) => $this->filterStatus0($q));
                $query->when($this->status === 1, fn(Builder $q) => $this->filterStatus1($q));
};

Project::tree()
    ->with([... some not recursive relations ...])
            ->withExists([
        'children',
    ])
    ->where($constraint)
    ->orWhereHas('children', $constraint)
    ->orWhereHas('children.children', $constraint)
    ->orWhereHas('children.children.children', $constraint)
    // ...
    ->get();
```

If the maximum depth can change, you could also build this query dynamically.

**Reply by paladam1** [reply]
Unfortunately no, it doesn't have, as it's a project tree with n sublevels that may vary by their types.

Thank you for your help, I'll try to figure out a solution or use another method to bypass this problem.
